### ISSUE TITLE
📟 Embedded Room    (at RustFest 2016)

### ISSUE THREAD

**Issue by gnunicorn** [RustFestEU/conf-2016#1]
This issue is meant the primary discussion point on what to do/work on in the **📟 Embedded Room**: an obvious playing feel for Rust are embedded devices and the Internet of Things. This room is dedicated to this topic.

If you have any material ready, are working on some (and/or want help with that), please post it here. If you just have a general idea (but no intention of creating it), please indicate that.

**If you plan to join this room, please indicate so by clicking the 👍-button at the end of this first post!**

**Comment by klingtnet** [comment]
I haven't done embedded programming besides some LED blinking but I would really like to get started with it. Should participants bring their own devices, like Arduinos or Raspberry Pi's?

**Comment by klingtnet** [comment]
An idea would be to program a super simple sound generator :loud_sound: , at least this is what I would like to create anyway.

**Comment by TorstenScheck** [comment]
I'm most interested in event-driven hierarchical state machines for embedded control software.
The mio or futures_mio crates would be fine for generating events from and to CAN and GPIO, I guess.
Then I want to process these events with a UML state machine similar to Qt's QStateMachine framework. It should be possible to use the resulting lib crates on bare-metal projects, but my main target would be Linux. So Rust integration into Poky/Yocto is more important to me than bare-metal challenges. Obviously, the "embedded" field is very broad...

But for the workshop I also don't mind trying to have some LEDs blink. :-)

**Comment by skade** [comment]
Well, the Workshops are intended to be meeting grounds instead of one person talking, many people listening.

RustFest should be the place where you meet people interested in your stuff, so hierarchical state machines is a great topic!

**Comment by SimonSapin** [comment]
I’ll be there, but leaving in the afternoon for the airport.

I’ve played and written about Rust on:
- BBC micro:bit https://github.com/SimonSapin/rust-on-bbc-microbit#readme
- Teensy 3 https://github.com/SimonSapin/teensy-clock#readme

I’ll bring the hardware for both.

**Comment by SimonSapin** [comment]
@klingtnet I don’t think anyone is gonna be handing out hardware, so if you want to play with some you should bring it :)

If you want to buy an Arduino board, be careful when choosing it as most of them have AVR microcontrollers. While https://github.com/avr-llvm/llvm and https://github.com/avr-rust/rust exist they seem fiddly to compile and use. You’ll make your life easier by picking one of the Arduino boards with an ARM processor.

Raspberry Pi’s typically run Linux, so they’re probably much easier to use with Rust than anything without an OS.

**Comment by klingtnet** [comment]
@SimonSapin Good to know that working with AVR's can be a bit fiddly, I was about to buy an Arduino Mega clone with such a controller.

**Comment by pwrdwnsys** [comment]
@SimonSapin Thanks for the links to your articles. I have ordered a micro:bit kit to use during the workshop.

**Comment by jamesmunns** [comment]
@SimonSapin I have two Teensy 3.2s (and might have a couple more at work I can borrow), I'll bring the ones I can.

**Comment by skade** [comment]
I got hold of the tessel 2 fleet from NodeSchool (10 pieces). They are labeled and named after those lables.

**Comment by mbr** [comment]
I'll bring a bare metal ARM MCU that runs a Rust firmware already and a thesis about it.=)

**Comment by vladimir-lu** [comment]
I'll be going to the embedded workshop with some arm hardware and lots of optimism... 😄

**Comment by jupp0r** [comment]
@TorstenScheck would you be interesting in starting to hack on something like Boost.MSM in Rust?

**Comment by TorstenScheck** [comment]
@jupp0r Yes, I plan to work on a UML state machine in rust. I need to wrap my head around an appropriate architecture though. It should be event loop agnostic and allow idiomatic rust usage. I will have to study Tokio and Rotor to get a better feeling for such a library.

So far, I've used QStatemachine, which is also an UML Statemachine/Statechart framework:
- UML specification
  - http://www.omg.org/spec/UML/2.5/
- Qt State Machine Framework Introduction
  - http://qt-project.org/doc/qt-4.8/statemachine-api.html
